{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a843765b-1a31-45b1-a8b3-42249bbab27b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os\n",
    "AD_NAME=\"songmingyang\"\n",
    "encrypted_password=\"iWRsYqbwV4EJgJvU8QjLe00CptZc5jBVH3FMo5i6n9mVdOSoUurpyBTmst1Z\"\n",
    "# 构造代理地址\n",
    "new_proxy_address = f\"http://{AD_NAME}:{encrypted_password}@10.1.20.50:23128/\"\n",
    "\n",
    "# 设置环境变量\n",
    "os.environ['http_proxy'] = new_proxy_address\n",
    "os.environ['https_proxy'] = new_proxy_address\n",
    "os.environ['HTTP_PROXY'] = new_proxy_address\n",
    "os.environ['HTTPS_PROXY'] = new_proxy_address\n",
    "\n",
    "# os.environ['HF_ENDPOINT'] = 'https://hf-mirror.com'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "efd04e4a-19f2-47d7-9cd1-aed5cf277805",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "Fetching 17 files: 100%|██████████| 17/17 [47:53<00:00, 169.02s/it]\n",
      "/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/torchvision/datapoints/__init__.py:12: UserWarning: The torchvision.datapoints and torchvision.transforms.v2 namespaces are still Beta. While we do not expect major breaking changes, some APIs may still change according to user feedback. Please submit any feedback you may have in this issue: https://github.com/pytorch/vision/issues/6753, and you can also check out https://github.com/pytorch/vision/issues/7319 to learn more about the APIs that we suspect might involve future changes. You can silence this warning by calling torchvision.disable_beta_transforms_warning().\n",
      "  warnings.warn(_BETA_TRANSFORMS_WARNING)\n",
      "/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/torchvision/transforms/v2/__init__.py:54: UserWarning: The torchvision.datapoints and torchvision.transforms.v2 namespaces are still Beta. While we do not expect major breaking changes, some APIs may still change according to user feedback. Please submit any feedback you may have in this issue: https://github.com/pytorch/vision/issues/6753, and you can also check out https://github.com/pytorch/vision/issues/7319 to learn more about the APIs that we suspect might involve future changes. You can silence this warning by calling torchvision.disable_beta_transforms_warning().\n",
      "  warnings.warn(_BETA_TRANSFORMS_WARNING)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-21 17:18:04,203 - lmdeploy - \u001b[33mWARNING\u001b[0m - archs.py:53 - Fallback to pytorch engine because `/mnt/petrelfs/songmingyang/.cache/huggingface/hub/models--Qwen--Qwen2-VL-7B-Instruct/snapshots/a28a094eb66a9f2ac70eef346f040d8a79977472` not supported by turbomind engine.\n",
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3508, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_74996/4082223376.py\", line 6, in <module>\n",
      "    piple = pipeline(model_name)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/api.py\", line 85, in pipeline\n",
      "    return pipeline_class(model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/serve/vl_async_engine.py\", line 37, in __init__\n",
      "    super().__init__(model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/serve/async_engine.py\", line 286, in __init__\n",
      "    self._build_pytorch(model_path=model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/serve/async_engine.py\", line 347, in _build_pytorch\n",
      "    self.engine = Engine(model_path=model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/engine/engine.py\", line 149, in __init__\n",
      "    self.model_agent = build_model_agent(\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/engine/model_agent.py\", line 762, in build_model_agent\n",
      "    model_agent = BaseModelAgent(model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/engine/model_agent.py\", line 214, in __init__\n",
      "    self.patched_model = self._build_model(model_path,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/engine/model_agent.py\", line 241, in _build_model\n",
      "    patched_model = build_patched_model(self.model_config, device=device)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/torch/utils/_contextlib.py\", line 115, in decorate_context\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/models/patch.py\", line 201, in build_patched_model\n",
      "    return build_model_from_hf_config(model_config, dtype=dtype, device=device)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/models/patch.py\", line 191, in build_model_from_hf_config\n",
      "    model_cls = _get_model_class(model_config, module_map)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/models/patch.py\", line 175, in _get_model_class\n",
      "    module_cls = _class_from_qualname(qualname)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/models/patch.py\", line 55, in _class_from_qualname\n",
      "    mod = importlib.import_module(modname)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/importlib/__init__.py\", line 126, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1050, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1027, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 688, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 883, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 241, in _call_with_frames_removed\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/models/qwen2_vl.py\", line 15, in <module>\n",
      "    from lmdeploy.pytorch.nn.linear import (build_colwise_linear,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/nn/linear.py\", line 155, in <module>\n",
      "    class BlockedF8Linear(nn.Module):\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/lmdeploy/pytorch/nn/linear.py\", line 165, in BlockedF8Linear\n",
      "    fp8_dtype: torch.dtype = torch.float8_e4m3fn,\n",
      "AttributeError: module 'torch' has no attribute 'float8_e4m3fn'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 2105, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1396, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1287, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1140, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 1055, in format_exception_as_a_whole\n",
      "    frames.append(self.format_record(record))\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 955, in format_record\n",
      "    frame_info.lines, Colors, self.has_colors, lvals\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/IPython/core/ultratb.py\", line 778, in lines\n",
      "    return self._sd.lines\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/core.py\", line 734, in lines\n",
      "    pieces = self.included_pieces\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/core.py\", line 677, in included_pieces\n",
      "    scope_pieces = self.scope_pieces\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/core.py\", line 614, in scope_pieces\n",
      "    scope_start, scope_end = self.source.line_range(self.scope)\n",
      "  File \"/mnt/petrelfs/songmingyang/anaconda3/envs/tool_server/lib/python3.10/site-packages/stack_data/core.py\", line 178, in line_range\n",
      "    return line_range(self.asttext(), node)\n",
      "AttributeError: 'Source' object has no attribute 'asttext'\n"
     ]
    }
   ],
   "source": [
    "from lmdeploy import pipeline\n",
    "from lmdeploy.vl import load_image\n",
    "\n",
    "model_name = \"Qwen/Qwen2-VL-7B-Instruct\"\n",
    "# model_name = \"/mnt/petrelfs/songmingyang/quxiaoye/models/Qwen2-VL-7B-Instruct\"\n",
    "piple = pipeline(model_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce20733f-00ab-4b4f-ad45-e7c4f0f7a5de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tool_server",
   "language": "python",
   "name": "tool_server"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
